server/src/main/java/org/elasticsearch/env/NodeMetadata.java

```diff
@@ -31,67 +32,116 @@
+               "nodeId='" + nodeId + '\'' +
+               ", nodeVersion=" + nodeVersion +
+               '}';
+    }
+
+    public String nodeId() {
+        return nodeId;
     }
 
-    private static ObjectParser<Builder, Void> PARSER = new ObjectParser<>("node_meta_data", Builder::new);
+    public Version nodeVersion() {
+        return nodeVersion;
+    }
 
-    static {
-        PARSER.declareString(Builder::setNodeId, new ParseField(NODE_ID_KEY));
+    public NodeMetadata upgradeToCurrentVersion() {
+        return upgradeToVersion(Version.CURRENT);
     }
 
-    public String nodeId() {
-        return nodeId;
+    NodeMetadata upgradeToVersion(Version upgradeVersion) {
+        if (nodeVersion.equals(Version.V_EMPTY)) {
+            assert upgradeVersion.onOrAfter(Version.V_4_6_0) : "version is required in the node metadata from CrateDB 4.6 onwards";
+            return new NodeMetadata(nodeId, upgradeVersion);
+        }
+
+        if (nodeVersion.before(upgradeVersion.minimumIndexCompatibilityVersion())) {
+            throw new IllegalStateException(
+                "cannot upgrade a node from version [" + nodeVersion + "] directly to version [" + upgradeVersion + "]");
+        }
+
+        //Ignore hotfix versions, because downgrades for hotfixes are allowed
+        if (upgradeVersion.beforeMajorMinor(nodeVersion)) {
+            throw new IllegalStateException(
+                "cannot downgrade a node from version [" + nodeVersion + "] to version [" + upgradeVersion + "]");
+        }
+
+        return nodeVersion.equals(upgradeVersion) ? this : new NodeMetadata(nodeId, upgradeVersion);
     }
 
     private static class Builder {
         String nodeId;
+        Version nodeVersion;
 
         public void setNodeId(String nodeId) {
             this.nodeId = nodeId;
         }
 
+        public void setNodeVersionId(int nodeVersionId) {
+            this.nodeVersion = Version.fromId(nodeVersionId);
+        }
+
         public NodeMetadata build() {
-            return new NodeMetadata(nodeId);
+            final Version nodeVersion;
+            if (this.nodeVersion == null) {
+                assert Version.CURRENT.onOrAfter(Version.V_4_6_0) : "version is required in the node metadata from CrateDB 4.6 onwards";
+                nodeVersion = Version.V_EMPTY;
+            } else {
+                nodeVersion = this.nodeVersion;
+            }
+
+            return new NodeMetadata(nodeId, nodeVersion);
         }
     }
 
+    static class NodeMetadataStateFormat extends MetadataStateFormat<NodeMetadata> {
+
+        private ObjectParser<Builder, Void> objectParser;
 
-    public static final MetadataStateFormat<NodeMetadata> FORMAT = new MetadataStateFormat<NodeMetadata>("node-") {
+        /**
+         * @param ignoreUnknownFields whether to ignore unknown fields or not. Normally we are strict about this, but
+         *                            {@link OverrideNodeVersionCommand} is lenient.
```

Comment:
Seem like I do not understand the use-case of `ignoreUnknownFields` now can I find this at the upsteam commits.
In which case do we set it to true? I could not see any usages.
I also do not understand in which case this data has unknown fields or in which case we want to ignore it.

Reply:
https://github.com/elastic/elasticsearch/blob/22ba759e1f37ef2607674074973a554f23730b8f/server/src/main/java/org/elasticsearch/env/NodeMetaData.java#L131

Reply:
This pr is based on the changes from elastic/elasticsearch@22ba759 and elastic/elasticsearch@749135b

Reply:
I guess this is just to make sure the parser does not break on unexpected fields in the data.

Reply:
Hm so looks like the other commit superseded https://github.com/elastic/elasticsearch/commit/ee23968f0509fc8dc7ffb79b79341b2d4b3b199f ?
To me it looks like its not used now anymore inside the `OverrideNodeVersionCommand` .
Anyhow shouldn't we then add the related test-case as well? https://github.com/elastic/elasticsearch/commit/ee23968f0509fc8dc7ffb79b79341b2d4b3b199f#diff-286fa4a3f35e6080a0348c4daa7b8e7da916b8ac4deeb386d6292a182add6d7aR162

Reply:
The actual test is https://github.com/elastic/elasticsearch/blob/22ba759e1f37ef2607674074973a554f23730b8f/server/src/test/java/org/elasticsearch/env/OverrideNodeVersionCommandTests.java

Reply:
The history is this is the following. When i moved the metadata from disk to lucene i skipped the versioning because the pr was missing. This is pretty much now the follow up on this and spans over 3 different prs. The original pr where the versioning was introduced, our change for the downgrade and the move to the lucene storage.

Reply:
However, it can be safely removed.

Reply:
Although the elasticsearch code base at this point still has it.

Reply:
Seem like my suspicion that this flag is not used anywhere beside the test I mentioned (https://github.com/elastic/elasticsearch/commit/ee23968f0509fc8dc7ffb79b79341b2d4b3b199f#diff-286fa4a3f35e6080a0348c4daa7b8e7da916b8ac4deeb386d6292a182add6d7aR162) is right. So I think we really can safely remove it. Thanks for doing so!

Reply:
Yes, you were on the right track.